File: Companies/VMware/ContiguousArray.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class ContiguousArray
{
public:
  int findMaxLength(vector<int> &nums)
  {
    unordered_map<int, int> m{};
    m[0] = -1;
    int maxlen{0}, sum{0};
    for (int i{0}; i < nums.size(); i++)
    {
      sum += nums[i] ? 1 : -1;
      if (m.find(sum) != m.end())
      {
        maxlen = max(maxlen, i - m[sum]);
      }
      else
      {
        m[sum] = i;
      }
    }
    return maxlen;
  }
};

File: Companies/Google/ContiguousArray.java
package Companies.Google;

import java.util.HashMap;
import java.util.Map;

public class ContiguousArray {
    public int findMaxLength(int[] nums) {
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, -1);
        int maxlen = 0, sum = 0;
        for (int i = 0; i < nums.length; i++) {
            sum = sum + (nums[i] == 1 ? 1 : -1);
            if (map.containsKey(sum)) {
                maxlen = Math.max(maxlen, i - map.get(sum));
            } else {
                map.put(sum, i);
            }
        }
        return maxlen;
    }
}
